Packages/com.llealloo.audiolink/Runtime/Scripts/YtdlpPlayer.cs

```diff
@@ -0,0 +1,303 @@
+        public string PlaybackTimeFormatted()
+        {
+            if(VideoPlayer != null && VideoPlayer.length > 0)
+            {
+                float videoLengthSeconds = (float)VideoPlayer.length;
+                float currentVideoTime = (float)VideoPlayer.time;
+
+                if(videoLengthSeconds >= 3600)
+                    return $"{TimeSpan.FromSeconds(currentVideoTime).ToString(@"hh\:mm\:ss")} / {TimeSpan.FromSeconds(videoLengthSeconds).ToString(@"hh\:mm\:ss")}";
+                else
+                    return $"{TimeSpan.FromSeconds(currentVideoTime).ToString(@"mm\:ss")} / {TimeSpan.FromSeconds(videoLengthSeconds).ToString(@"mm\:ss")}";
+            }
+            else
+                return "00:00 / 00:00";
+        }
+    }
+
+    [CustomEditor(typeof(YtdlpPlayer))]
+    public class YtdlpPlayerCleanEditor : Editor 
+    {
+        YtdlpPlayer _ytdlpPlayer;
+        bool _reloadURL = false;
+
+        void OnEnable()
+        {
+            _ytdlpPlayer = (YtdlpPlayer) target;
+        }
+
+        //TODO: add a warning on Linux that only some filetypes are supported?
+        public override void OnInspectorGUI()
+        {
+            base.OnInspectorGUI();
+            float playbackTime = 0;
+            bool hasPlayer = _ytdlpPlayer.VideoPlayer != null;
+            if(hasPlayer && _ytdlpPlayer.VideoPlayer.length > 0)
+                playbackTime = _ytdlpPlayer.GetPlaybackTime();
+            
+            EditorGUI.BeginDisabledGroup(!hasPlayer || !Application.IsPlaying(target) || !_ytdlpPlayer.VideoPlayer.isPlaying);
+            _reloadURL = EditorGUILayout.Toggle("Reload URL", _reloadURL);
+            if(_reloadURL)
+            {
+                _ytdlpPlayer.UpdateAndPlay();
+                _reloadURL = false;
+            }
+
+            EditorGUI.BeginChangeCheck();
+            EditorGUILayout.LabelField(new GUIContent(" Seek: " + _ytdlpPlayer.PlaybackTimeFormatted(), EditorGUIUtility.IconContent("d_Slider Icon").image));
+            playbackTime = EditorGUILayout.Slider(playbackTime, 0, 1);
+            if(EditorGUI.EndChangeCheck())
+                _ytdlpPlayer.SetPlaybackTime(playbackTime);
+
+            EditorGUI.EndDisabledGroup();
+            
+        }
+    }
+
+    [InitializeOnLoad]
+    public static class YtdlpURLResolver
+    {
+        private static string _ytdlpDownloadURL = "https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp.exe";
+        private static string _localYtdlpPath = Application.dataPath + "\\AudioLink\\yt-dlp.exe";
+
+        private static string _ytdlpPath = "";
+        private static HashSet<Process> _runningYtdlProcesses = new HashSet<Process>();
+        private static bool _ytdlFound = false;
+
+        /// <summary> Initialize URL Resolver </summary>
+        static YtdlpURLResolver()
+        {
+            EditorApplication.playModeStateChanged += PlayModeChanged;
+            LocateYtdlp();
```

Comment:
This is pretty cursed, I'm not gonna lie. You have no control over static initialization, and in Unity especially, static fields act in very funky ways. It also means that you will potentially get nagged by the script even if you never plan to use it. The class should probably just be a lazy singleton if you absolutely need global state.

More pressingly, you have no way to handle cleanup. You spin up processes and put them into this hashset, but if a domain reload happens, the hashset will be reset, and the processes will be left dangling, unless I am missing something.

I don't really get what the point of keeping the processes in this hashmap is, anyways. Could you enlighten me? Is there ever more than 1 in here?